Switch pre-commit to prek (#54258)
mistercrunch/Airflow: contributing-docs/08_static_code_checks.rst

==================

The static code checks in Airflow are used to verify that the code meets certain quality standards.
All the static code checks can be run through prek hooks.

The prek hooks perform all the necessary installation when you run them
for the first time. See the table below to identify which prek checks require the Breeze Docker images.

You can also run the checks via `Breeze <../dev/breeze/doc/README.rst>`_ environment.

**The outline for this document in GitHub is available at top-right corner button (with 3-dots and 3 lines).**

Prek hooks
----------

> THe pre-commit hooks help speed up your local development cycle and place less burden on the CI infrastructure.

Reviewer: ```suggestion
The pre-commit hooks help speed up your local development cycle and place less burden on the CI infrastructure.
```

Author: Yep.

In such cases, the Docker-based prek will inform you that you should rebuild the image.

In case you do not have your local images built, the pre-commit hooks fail and provide
instructions on what needs to be done.
Enabling prek hooks
-------------------

Enabling pre-commit hooks
-------------------------

To turn on pre-commit checks for ``commit`` operations in git, enter:
To turn on prek checks for ``commit`` operations in git, enter:

.. code-block:: bash

    pre-commit install

    prek install

To install the checks also for ``pre-push`` operations, enter:

.. code-block:: bash

    pre-commit install -t pre-push

    prek install -t pre-push

For details on advanced usage of the install method, use:

.. code-block:: bash

   pre-commit install --help
   prek install --help

Available pre-commit checks
---------------------------
.. note::

    The ``prek`` tool is a drop-in replacement for the legacy ``pre-commit`` tool - much faster and more
    feature-rich, If you have already installed ``pre-commit`` to handle your hooks, you can run
    ``prek install -f`` to replace the existing ``pre-commit`` hooks with the ``prek`` hooks.

Using prek
----------

> This table lists pre-commit hooks used by Airflow. The ``Image`` column indicates which hooks

Reviewer 2: Is this table not required ? its still valid right?

Author: Yeah. It's a bit paint to keep it updated - always some conflicts to resolve, and since we have `prek list --verbose` we can very easily see the list of available pre-commits with description, and auto-complete has all those descriptions, I think it makes very little sense to keep the table here.

Author: I restored the "Available prek hooks" chapter - but I only described there how to get the list of available prek hooks.

Reviewer 2: Yeah make sense :)

And if you want to see the details of a particular hook, you can run:

.. code-block:: bash

    prek list --verbose <hook-id>

When you install auto-completion, you can also use the tab-completion to see the available hooks.

Using prek
----------

After installation, prek hooks are run automatically when you commit the
code. But you can run prek hooks manually as needed.

-   Run all checks on your staged files by using:

.. code-block:: bash

    prek

-   Run only mypy check on your staged airflow and dev files by specifying the
    ``mypy-airflow-core`` and ``mypy-dev`` prek hooks (more hooks can be specified):

.. code-block:: bash

    prek mypy-airflow-core mypy-dev

-   Run only mypy airflow checks on all "airflow-core" files by using:

.. code-block:: bash

    prek mypy-airflow-core --all-files

-   Run all checks on all files by using:

.. code-block:: bash

    prek --all-files

-   Run all checks only on files modified in the last locally available commit in your checked out branch:

.. code-block:: bash

    prek --last-commit

-   Run all checks only on files modified in your last branch that is targeted to be merged into the main branch:

.. code-block:: bash

>     prek --base-from main

Reviewer: ```suggestion
    prek --from-ref main
```

Author: Indeed :)

-   Show files modified automatically by prek when prek automatically fix errors

.. code-block:: bash

    prek --show-diff-on-failure

-   Skip one or more of the checks by specifying a comma-separated list of
    checks to skip in the SKIP variable:

.. code-block:: bash

    SKIP=mypy-airflow-core,ruff prek --all-files

You can always skip running the tests by providing ``--no-verify`` flag to the
``git commit`` command.

> To check other usage types of the prek framework, see `prek website <https://prek.com/>`__.

Reviewer: ```suggestion
To check other usage types of the pre-commit framework, see `Pre-commit website <https://pre-commit.com/>`__.
```

Author: Yeah.

You can add the variable to your ``.bashrc`` or ``.zshrc`` if you do not want to set it manually every time you
enter the terminal.

Manual prek hooks
-----------------

Most of the checks we run are configured to run automatically when you commit the code. However,
there are some checks that are not run automatically and you need to run them manually. Those
checks are marked with ``manual`` in the ``Description`` column in the table below. You can run
them manually by running ``prek run --hook-stage manual <hook-id>``.

Special pin-versions prek

> -------------------------------

Reviewer 2: ```suggestion
Special pin-versions prek
-------------------------
```

will handle refreshing the cache when and if needed, there are some cases when it won't (cache invalidation
is the hard problem in computer science). This might happen for example when we upgrade MyPY. In such
cases you might need to manually remove the cache volume by running ``breeze down --cleanup-mypy-cache``.

Debugging prek check scripts requiring image
--------------------------------------------------

Those commits that use Breeze docker image might sometimes fail, depending on your operating system and
docker setup, so sometimes it might be required to run debugging with the commands. This is done via
two environment variables ``VERBOSE`` and ``DRY_RUN``. Setting them to "true" will respectively show the
commands to run before running them or skip running the commands.

Note that you need to run prek with --verbose command to get the output regardless of the status
of the static check (normally it will only show output on failure).

Printing the commands while executing:

.. code-block:: bash

     VERBOSE="true" prek --verbose ruff

Just performing dry run:

.. code-block:: bash

>      DRY_RUN="true" prek --verbose ruff

Reviewer: `VERBOSE` and `DRY_RUN` don't work with prek for now, and I doubt them ever worked in `pre-commit` either(I can't find them in the docs or code). Do they comes from the `breeze static-checks` wrapper?

Author: Yes - they are breeze-specific variables. This is quite "deep" and already largely not-needed chapter. I removed it entirely.
